server/text_generation_server/utils/weights.py

```diff
@@ -9,6 +10,93 @@
 from text_generation_server.utils.log import log_once
 
 
+@dataclass
+class GPTQWeight:
+    qweight: torch.Tensor
+    qzeros: torch.Tensor
+    scales: torch.Tensor
+    g_idx: Optional[torch.Tensor]
+    bits: int
+    groupsize: int
+    use_exllama: bool
+
+    # These are generated by the Exllama linear layer, but we probably
+    # need to keep them in Python-land.
+    q_perm: Optional[torch.Tensor] = None
+    q_invperm: Optional[torch.Tensor] = None
+
+    @property
+    def device(self) -> torch.device:
+        return self.qweight.device
+
+
+@dataclass
+class Exl2Weight:
+    """
+    Exllama2 exl2 quantized weights.
+    """
+
+    q_weight: torch.Tensor
+    q_scale: torch.Tensor
+    q_invperm: torch.Tensor
+    q_perm: torch.Tensor
+    q_scale_max: torch.Tensor
+    q_groups: torch.Tensor
+    q_group_map: torch.Tensor
+
+    @classmethod
+    def get_tensor(cls, prefix, weight: "Weights"):
+        tensors = {}
+        for key in Exl2Weight._load_tensor_keys():
+            # TODO: exl2 sharding.
+            name = f"{prefix}.{key}"
+            try:
+                tensors[key] = weight.get_tensor(name)
+            except RuntimeError:
+                raise RuntimeError(
+                    f"Cannot load `exl2`-quantized weight {name}, make sure the model is already quantized."
+                )
+
+        tensors["q_group_map"] = Exl2Weight.make_group_map(
+            tensors["q_groups"], tensors["q_weight"].shape[0]
+        )
+        tensors["q_perm"] = torch.argsort(tensors["q_invperm"]).to(torch.int)
+
+        w = Exl2Weight(**tensors)
+        w.q_scale_max /= 256
+        w.q_perm = w.q_perm.short()
+        w.q_invperm = w.q_invperm.short()
+
+        return w
+
+    @staticmethod
+    def make_group_map(q_groups: torch.Tensor, num_qrows: int):
```

Comment:
This function actually already exists in `exllamav2.py`  let's not rewrite it.

Reply:
I had a few comments about it, but since it's coming from external repo (which should be indicated somewhere). It's not worth changing anything, the benefit of leaving as-is is that we can just pull updates whenever they come.

Reply:
Moved back to `exllamav2.py`